### DISCUSSION TITLE
Is there something like dynamic routing and routing based on traffic history (time-dependent graph/dynamic graph)?

### DISCUSSION THREAD

**Discussion by ldsmoreira** [valhalla/valhalla#4030] [General]
Hello Valhalla community!

I have a question related to route optimization based on historical traffic data. I'm curious to know if there is an existing feature or plan in the roadmap that enables the calculation of optimized routes by taking into account **traffic history data** and the time spent on previously traveled paths in order to determine the next best path. I believe this concept might be associated with **dynamic graphs**/**time-dependent graphs**.

If a feature like this already exists, or if there is interest in developing it, I'd be more than happy to contribute to its implementation or enhancement. Thank you!

**Comment by nilsnolde** [comment]
Jep, we call it mostly predicted traffic and it’s fully functional already. On mobile right now, so linking is tedious, but I even wrote one or two articles on the topic (IIRC😅). Just look for „predicted“ in the codebase and you shall get lucky:)

**Reply by nilsnolde** [reply]
Though I can’t claim any of the idea, it’s been there before I arrived.

**Reply by ldsmoreira** [reply]
Hi there! I appreciate the swift response!

My ultimate goal is to create a custom, adaptive costing model based on datasets generated from client data. At this stage, I'm just trying to validate the concept before diving into actual implementation. Although this is currently a one-person experiment, I'm excited about the learning experience and the fun that comes along with it!

I'm seeking suggestions on where to begin – perhaps something like the "Hello World" equivalent for custom costing models. As I'm going through the documentation, I'm considering adding some "print" statements within the Sif code to gain a better understanding of its inner workings. Additionally, I believe that familiarizing myself with Valhalla map tiles could prove beneficial.

What are your thoughts on this approach? Any guidance or suggestions on where to start would be greatly appreciated. Thank you!

**Reply by nilsnolde** [reply]
Hm it depends a bit what you mean with „adaptive costing model“ and what the „client datasets” are exactly. You mean historical traffic data generated from client data? Or other datasets?

**Reply by ldsmoreira** [reply]
I mean historical traffic data generated by clients devices. So the dataset would grow every day and i could do things like train a model to predict the cost of a given edge at a given time (example).

**Reply by kevinkreiser** [reply]
we have no tooling readily available to support the gathering and production of traffic data. we have another repository that is related to that effort: https://github.com/OpenStreetMapSpeeds/conflation

this can turn some gps data into speeds on roads. its not current time of day aware but we could easily make it so. you'd also want to add a new input to it for your own data pipeline.

once you had that, and the code to generate valhalla style historical data from it (examples of this exist in unit tests) then you would need to periodically update your tileset with this updated data and restart your routing service to make the changes live.

since it is historical data i think doing it weekly is probably the absolute maximum frequency you'd want to run this to get updated speeds.

anyway its a lot of work but it is doable for sure!

**Reply by ldsmoreira** [reply]
Ohh, good!

I was thinking that i may need to give my first steps in valhalla firstly and then put my hands on conflation. If i go straight to conflation maybe i will be lost soon hahaha!

I'm already trying to understand basic stuff, such as how tiles are parsed, how valhalla represents graph tile stuff in memory, some of the algorithms, and so on.

I wil try to follow the request for a route from the beggining to output putting some probes here and there to be able to get the general picture of the engine.

Do you know some path that i may follow that will make me get a good understand of valhalla's internals ?

Obs: I'm pretty new in valhalla, but my goal is to contribute to the project in the future as well as use valhalla to my own projects and ventures.

**Reply by kevinkreiser** [reply]
you can watch how a request is handled by reviewing the `actions` that the `actor_t` class provides. it shows all the steps for request handling. for data processing you'll need to review `mjolnir` its the thing that turns data sources into tiles we can route on (service requests with).

**Comment by ldsmoreira** [comment]
Hello again!

As I continue to study Valhalla, I'm trying to come up with a simple yet doable proof of concept to work on, and I'd appreciate some advice on whether my current approach is valuable or if I might be heading in the wrong direction.

My PoC idea is straightforward:

1. Receive an updated speed for a given edge/way.
2. Update the speed in the tile.

I understand that this approach would require persistently updating the speeds of tiles, but I think it might work reasonably well for the main roads in Rio de Janeiro, where I live, as an example. The roads could be updated on a regular basis.

The end goal of this experiment would be to develop a new service that listens for speed changes and triggers the reassignment of a speed to a given road.

What are your thoughts on this approach? Are there any recommendations you'd like to share? I'm entirely receptive to your insights to help steer a proof of concept in this direction!

Obs: It's important to note that this PoC doesn't focus on historical data. However, working on this project should provide valuable insights into the characteristics of the data itself and help me better understand the overall data processing flow.

**Reply by ldsmoreira** [reply]
I came across the valhalla_add_predicted_traffic as @nilsnolde  mentioned, and it appears to be quite similar to the Proof of Concept I initially proposed. Do you have any suggestions or directions that could help me better understand and potentially contribute to the improvement of Valhalla in this subject [in the future] (e.g, historical data, dynamic speed update, ...)?

**Reply by kevinkreiser** [reply]
we do in fact already support both of these things. the thing we don't offer in the repo though are the operational tools for the data pipelines to get your:

1. historical data into the format it needs to be to be added to the tiles
2. real time data continually updated in the traffic tar

the reason we don't offer these is simply because there's no sensible way to do it that would work for everyone's needs. different people have different datasources and the formats, cadence, means of delivery coupled with the shape of thier deployment makes it very difficult to offer something that would work for everyone.

that said I did mean to write up an issue or Readme explaining how it's supposed to work in general so that each person could build thier own operational scripting particular to their situation.

tldr; you can think of this stuff more as operational where it doesn't make sense for the project to offer strong opinions on that

**Reply by nilsnolde** [reply]
I'll have to drop the shameless plug here: we already solved some of what you want @ldsmoreira and offer it as a commercial service and/or product. Feel free to contact me via email (public Github email) if you want to discuss around that.

**Comment by Trietes** [comment]
For anything regarding live and historical traffic I can also reccomend the POC of alinmindroc: https://github.com/alinmindroc/valhalla_traffic_poc

**Reply by Jofemago** [reply]
But here is a problem, how we built the traffic.rar, using traffic from here or tomtom, and how we get the edges and nodes with a specific location (latitude, longitude)

**Reply by Trietes** [reply]
Getting a source of traffic data and using it in Valhalla is a different problem. The traffic data from Mapbox is already mapped to OSM segements. If you want to use HERE or TomTom traffic data, you first have to see if they allow using their traffic data on OSM data. Or you have to route with valhalla on their card material. If you want to do that, you might want to contact @nilsnolde

**Reply by Jofemago** [reply]
Really, thank you very much for the information. It's not that I want to use Here or Valhalla, I was looking for a more convenient alternative to incorporate traffic into Valhalla. Thank you very much for opening my eyes and giving me more hints on where to look. One thing that is not clear to me is that the documentation mentions a traffic.rar file, but it doesn't explain what should be inside that file. I want to conduct tests with an entire city, like Bogotá, Colombia, which is ranked among the worst in terms of traffic worldwide. Additionally, I believe that Mapbox isn't as bad at calculating distances compared to Google Traffic, nor is its traffic data, as shown in the following distance graph. It illustrates the difference between Google and traffic data from Mapbox: Google without traffic data compared to Mapbox with traffic data. Similarly, this holds true for Valhalla and OSRM, both of which are well-known open-source tools. This is particularly noticeable in terms of distances, where they provide very accurate values for use in open-source applications.

# average error in metters(mae) google with trafficc- another api
![image](https://github.com/valhalla/valhalla/assets/18221796/49706c26-6e88-41c7-aa31-55ab30518001)

I also want to share with you the time comparison. Here, it's important to highlight that Mapbox's performance, compared to Google with traffic (which represents the traffic we aim to achieve, or rather, is used as the ground truth for traffic data), is quite impressive. Even Google without traffic data provides good results. However, since traffic data wasn't included for Valhalla and OSRM, their time estimates are not very reliable. In conclusion, it's highly advantageous that Mapbox offers a compatible format. This could yield excellent results with proper integration.

# average error in seconds(mae) google with trafficc- another api
![image](https://github.com/valhalla/valhalla/assets/18221796/dd10c38c-5f6d-4226-98bc-418fae65de3b)

For these methods, 50 points were calculated, and the distance and time between each point were determined. Tables were constructed, and an Exploratory Data Analysis (EDA) of the data was performed. The graphs were generated using Seaborn in Python.

**Comment by nilsnolde** [comment]
Live traffic can usually be used on OSM data (or any other data, that’s what openlr is for). Historical is different, but probably just a negotiation item. To do this at scale is not the most trivial problem.
